File: sdks/typescript/chio-ts/test/manifest.test.ts
import test from "node:test";
import assert from "node:assert/strict";

import {
  signJsonStringEd25519,
  verifySignedManifest,
  verifySignedManifestJson,
} from "../src/index.ts";
import type { SignedManifest } from "../src/index.ts";

function pricedSignedManifest(): SignedManifest {
  return {
    manifest: {
      schema: "chio.manifest.v1",
      server_id: "srv-priced",
      name: "Priced Server",
      version: "1.0.0",
      tools: [
        {
          name: "greet",
          description: "Returns a greeting",
          input_schema: { type: "object" },
          pricing: {
            pricing_model: "per_invocation",
            unit_price: { units: 25, currency: "USD" },
            billing_unit: "invocation",
          },
          has_side_effects: false,
          latency_hint: "instant",
        },
      ],
      public_key: "22".repeat(32),
    },
    signature: "33".repeat(64),
    signer_key: "11".repeat(32),
  };
}

test("signed manifest typing and verification preserve pricing metadata", () => {
  const signedManifest = pricedSignedManifest();

  assert.equal(
    signedManifest.manifest.tools[0].pricing?.unit_price?.units,
    25,
  );

  const verification = verifySignedManifest(signedManifest);
  assert.equal(verification.structure_valid, true);
  assert.equal(verification.signature_valid, false);
  assert.equal(verification.embedded_public_key_valid, true);
  assert.equal(verification.embedded_public_key_matches_signer, false);
});

test("manifest structure does not include embedded public key validity", () => {
  const signedManifest = pricedSignedManifest();
  signedManifest.manifest.public_key = "demo-placeholder";

  const verification = verifySignedManifest(signedManifest);

  assert.equal(verification.structure_valid, true);
  assert.equal(verification.embedded_public_key_valid, false);
  assert.equal(verification.embedded_public_key_matches_signer, false);
});

test("manifest structure rejects empty or padded identity fields", () => {
  for (const [field, value] of [
    ["server_id", ""],
    ["server_id", " srv-priced"],
    ["server_id", "srv-priced "],
    ["server_id", "srv\npriced"],
    ["name", ""],
    ["name", " Priced Server"],
    ["name", "Priced Server "],
    ["version", ""],
    ["version", " 1.0.0"],
    ["version", "1.0.0 "],
  ] as const) {
    const signedManifest = pricedSignedManifest();
    signedManifest.manifest[field] = value;

    const verification = verifySignedManifest(signedManifest);

    assert.equal(
      verification.structure_valid,
      false,
      `${field} ${JSON.stringify(value)}`,
    );
    assert.equal(verification.signature_valid, false);
    assert.equal(verification.embedded_public_key_valid, true);
  }
});

test("manifest structure rejects empty or padded tool names", () => {
  for (const name of ["", " greet", "greet "]) {
    const signedManifest = pricedSignedManifest();
    signedManifest.manifest.tools[0].name = name;

    const verification = verifySignedManifest(signedManifest);

    assert.equal(
      verification.structure_valid,
      false,
      `name ${JSON.stringify(name)}`,
    );
  }
});

test("manifest structure rejects non-object tool schemas", () => {
  const badInputSchema = pricedSignedManifest();
  badInputSchema.manifest.tools[0].input_schema = [];
  assert.equal(verifySignedManifest(badInputSchema).structure_valid, false);

  const badOutputSchema = pricedSignedManifest();
  badOutputSchema.manifest.tools[0].output_schema = "not an object";
  assert.equal(verifySignedManifest(badOutputSchema).structure_valid, false);
});

test("manifest structure rejects non-object tool entries", () => {
  const signedManifest = pricedSignedManifest();
  (signedManifest.manifest.tools as unknown[])[0] = null;

  assert.equal(verifySignedManifest(signedManifest).structure_valid, false);
});

test("manifest structure rejects malformed pricing metadata", () => {
  for (const [label, pricing] of [
    [
      "per_invocation missing unit_price",
      { pricing_model: "per_invocation", billing_unit: "invocation" },
    ],
    [
      "per_unit missing billing_unit",
      { pricing_model: "per_unit", unit_price: { units: 25, currency: "USD" } },
    ],
    [
      "hybrid missing base_price",
      {
        pricing_model: "hybrid",
        unit_price: { units: 25, currency: "USD" },
        billing_unit: "document",
      },
    ],
    [
      "padded billing_unit",
      {
        pricing_model: "per_invocation",
        unit_price: { units: 25, currency: "USD" },
        billing_unit: " invocation",
      },
    ],
    [
      "invalid currency",
      {
        pricing_model: "per_invocation",
        unit_price: { units: 25, currency: "usd" },
        billing_unit: "invocation",
      },
    ],
    [
      "units above u64 max",
      {
        pricing_model: "per_invocation",
        unit_price: { units: 2 ** 64, currency: "USD" },
        billing_unit: "invocation",
      },
    ],
  ] as const) {
    const signedManifest = pricedSignedManifest();
    (signedManifest.manifest.tools[0] as { pricing?: unknown }).pricing = pricing;

    assert.equal(
      verifySignedManifest(signedManifest).structure_valid,
      false,
      label,
    );
  }
});

test("manifest structure accepts Rust u64 pricing units above JS safe integer", () => {
  const signedManifest = pricedSignedManifest();
  const largeRustU64Units = Number.MAX_SAFE_INTEGER + 2;
  assert.equal(Number.isSafeInteger(largeRustU64Units), false);
  (signedManifest.manifest.tools[0] as { pricing?: { unit_price?: { units?: number } } })
    .pricing!.unit_price!.units = largeRustU64Units;

  assert.equal(verifySignedManifest(signedManifest).structure_valid, true);
});

test("signed manifest envelope rejects unknown top-level fields", () => {
  const signedManifest = pricedSignedManifest() as SignedManifest & {
    unsigned_policy_hint?: unknown;
  };
  signedManifest.unsigned_policy_hint = { allow: true };

  assert.equal(verifySignedManifest(signedManifest).structure_valid, false);
});

test("manifest structure accepts valid required permissions", () => {
  const signedManifest = pricedSignedManifest();
  signedManifest.manifest.required_permissions = {
    read_paths: ["/tmp/in"],
    write_paths: ["/tmp/out"],
    network_hosts: ["api.example.com"],
    environment_variables: ["TOKEN"],
  };

  const verification = verifySignedManifest(signedManifest);

  assert.equal(verification.structure_valid, true);
  assert.equal(verification.signature_valid, false);
  assert.equal(verification.embedded_public_key_valid, true);
});

test("manifest structure rejects invalid required permissions", () => {
  for (const [field, values] of [
    ["read_paths", [""]],
    ["write_paths", [" /tmp/out"]],
    ["network_hosts", ["api.example.com "]],
    ["environment_variables", ["TOKEN", "TOKEN"]],
    ["read_paths", ["/tmp/in\nbad"]],
    ["read_paths", [123]],
  ] as const) {
    const signedManifest = pricedSignedManifest();
    signedManifest.manifest.required_permissions = { [field]: values };

    const verification = verifySignedManifest(signedManifest);

    assert.equal(
      verification.structure_valid,
      false,
      `${field} ${JSON.stringify(values)}`,
    );
    assert.equal(verification.signature_valid, false);
    assert.equal(verification.embedded_public_key_valid, true);
  }
});

test("manifest structure rejects malformed required permissions object", () => {
  const unknownField = pricedSignedManifest();
  (unknownField.manifest as { required_permissions?: unknown }).required_permissions = {
    unknown: ["/tmp"],
  };
  assert.equal(verifySignedManifest(unknownField).structure_valid, false);

  const nonArrayValues = pricedSignedManifest();
  (nonArrayValues.manifest as { required_permissions?: unknown }).required_permissions = {
    read_paths: "/tmp",
  };
  assert.equal(verifySignedManifest(nonArrayValues).structure_valid, false);
});

test("signed manifest JSON verification is fail-soft for malformed envelopes", () => {
  for (const input of [
    "{}",
    "{\"manifest\":null,\"signature\":\"\",\"signer_key\":\"\"}",
  ]) {
    const verification = verifySignedManifestJson(input);

    assert.equal(verification.structure_valid, false);
    assert.equal(verification.signature_valid, false);
    assert.equal(verification.embedded_public_key_valid, false);
    assert.equal(verification.embedded_public_key_matches_signer, false);
  }
});

test("manifest structure rejects missing required tool fields and nested unknown fields", () => {
  const missingDescription = pricedSignedManifest();
  delete (missingDescription.manifest.tools[0] as { description?: unknown }).description;
  assert.equal(verifySignedManifest(missingDescription).structure_valid, false);

  const missingSideEffects = pricedSignedManifest();
  delete (missingSideEffects.manifest.tools[0] as { has_side_effects?: unknown }).has_side_effects;
  assert.equal(verifySignedManifest(missingSideEffects).structure_valid, false);

  const unknownManifestField = pricedSignedManifest();
  (unknownManifestField.manifest as { unsigned_policy_hint?: unknown }).unsigned_policy_hint = true;
  assert.equal(verifySignedManifest(unknownManifestField).structure_valid, false);

  const unknownToolField = pricedSignedManifest();
  (unknownToolField.manifest.tools[0] as { annotations?: unknown }).annotations = {};
  assert.equal(verifySignedManifest(unknownToolField).structure_valid, false);

  const unknownPricingField = pricedSignedManifest();
  (
    unknownPricingField.manifest.tools[0].pricing as {
      experimental_discount?: unknown;
    }
  ).experimental_discount = 10;
  assert.equal(verifySignedManifest(unknownPricingField).structure_valid, false);
});

test("manifest structure validates server_tools and latency_hint enum values", () => {
  const valid = pricedSignedManifest();
  (valid.manifest as { server_tools?: string[] }).server_tools = ["bash", "text_editor"];
  valid.manifest.tools[0].latency_hint = "fast";
  assert.equal(verifySignedManifest(valid).structure_valid, true);

  const duplicateServerTool = pricedSignedManifest();
  (duplicateServerTool.manifest as { server_tools?: string[] }).server_tools = ["bash", "bash"];
  assert.equal(verifySignedManifest(duplicateServerTool).structure_valid, false);

  const unknownServerTool = pricedSignedManifest();
  (unknownServerTool.manifest as { server_tools?: string[] }).server_tools = ["database"];
  assert.equal(verifySignedManifest(unknownServerTool).structure_valid, false);

  const invalidLatencyHint = pricedSignedManifest();
  (invalidLatencyHint.manifest.tools[0] as { latency_hint?: string }).latency_hint = "immediate";
  assert.equal(verifySignedManifest(invalidLatencyHint).structure_valid, false);
});

test("signed manifest JSON verification preserves raw large integer tokens", () => {
  const seedHex = "01".repeat(32);
  const publicKeyHex = signJsonStringEd25519("{}", seedHex).public_key_hex;
  const rawManifest =
    `{"schema":"chio.manifest.v1","server_id":"srv-large-u64","name":"Large U64","version":"1.0.0","tools":[{"name":"price","description":"Returns price","input_schema":{"type":"object"},"pricing":{"pricing_model":"per_invocation","unit_price":{"units":9223372036854775808,"currency":"USD"},"billing_unit":"invocation"},"has_side_effects":false}],"public_key":"${publicKeyHex}"}`;
  const signed = signJsonStringEd25519(rawManifest, seedHex);
  const signedManifestJson =
    `{"manifest":${rawManifest},"signature":"${signed.signature_hex}","signer_key":"${signed.public_key_hex}"}`;

  const jsonVerification = verifySignedManifestJson(signedManifestJson);
  assert.equal(jsonVerification.structure_valid, true);
  assert.equal(jsonVerification.signature_valid, true);
  assert.equal(jsonVerification.embedded_public_key_valid, true);
  assert.equal(jsonVerification.embedded_public_key_matches_signer, true);

  const objectVerification = verifySignedManifest(JSON.parse(signedManifestJson));
  assert.equal(objectVerification.structure_valid, true);
  assert.equal(objectVerification.signature_valid, false);
});

File: sdks/go/chio-go/invariants/manifest_test.go
package invariants_test

import (
	"encoding/json"
	"strings"
	"testing"

	"github.com/backbay-labs/chio/sdks/go/chio-go/invariants"
)

func pricedSignedManifest() map[string]any {
	return map[string]any{
		"manifest": map[string]any{
			"schema":    "chio.manifest.v1",
			"server_id": "srv-priced",
			"name":      "Priced Server",
			"version":   "1.0.0",
			"tools": []any{
				map[string]any{
					"name":         "greet",
					"description":  "Returns a greeting",
					"input_schema": map[string]any{"type": "object"},
					"pricing": map[string]any{
						"pricing_model": "per_invocation",
						"unit_price": map[string]any{
							"units":    float64(25),
							"currency": "USD",
						},
						"billing_unit": "invocation",
					},
					"has_side_effects": false,
				},
			},
			"public_key": strings.Repeat("22", 32),
		},
		"signature":  strings.Repeat("33", 64),
		"signer_key": strings.Repeat("11", 32),
	}
}

func TestManifestStructureDoesNotIncludeEmbeddedPublicKeyValidity(t *testing.T) {
	signedManifest := pricedSignedManifest()
	signedManifest["manifest"].(map[string]any)["public_key"] = "demo-placeholder"

	verification, err := invariants.VerifySignedManifest(signedManifest)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error: %v", err)
	}
	if !verification.StructureValid {
		t.Fatalf("malformed embedded public key must not invalidate manifest structure")
	}
	if verification.EmbeddedPublicKeyValid {
		t.Fatalf("malformed embedded public key must be reported separately")
	}
	if verification.EmbeddedPublicKeyMatchesSigner {
		t.Fatalf("malformed embedded public key must not match signer")
	}
}

func TestManifestStructureRejectsEmptyOrPaddedIdentityFields(t *testing.T) {
	cases := []struct {
		field string
		value string
	}{
		{"server_id", ""},
		{"server_id", " srv-priced"},
		{"server_id", "srv-priced "},
		{"server_id", "srv\npriced"},
		{"name", ""},
		{"name", " Priced Server"},
		{"name", "Priced Server "},
		{"version", ""},
		{"version", " 1.0.0"},
		{"version", "1.0.0 "},
	}
	for _, tc := range cases {
		t.Run(tc.field+"="+tc.value, func(t *testing.T) {
			signedManifest := pricedSignedManifest()
			signedManifest["manifest"].(map[string]any)[tc.field] = tc.value

			verification, err := invariants.VerifySignedManifest(signedManifest)
			if err != nil {
				t.Fatalf("VerifySignedManifest returned error: %v", err)
			}
			if verification.StructureValid {
				t.Fatalf("%s value %q must be structurally invalid", tc.field, tc.value)
			}
			if verification.SignatureValid {
				t.Fatalf("mutated identity field should invalidate the signature")
			}
			if !verification.EmbeddedPublicKeyValid {
				t.Fatalf("embedded public key validity must remain independent")
			}
		})
	}
}

func TestManifestStructureRejectsEmptyOrPaddedToolNames(t *testing.T) {
	for _, name := range []string{"", " greet", "greet "} {
		t.Run(name, func(t *testing.T) {
			signedManifest := pricedSignedManifest()
			tool := signedManifest["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)
			tool["name"] = name

			verification, err := invariants.VerifySignedManifest(signedManifest)
			if err != nil {
				t.Fatalf("VerifySignedManifest returned error: %v", err)
			}
			if verification.StructureValid {
				t.Fatalf("name %q must be structurally invalid", name)
			}
		})
	}
}

func TestManifestStructureRejectsNonObjectToolSchemas(t *testing.T) {
	badInputSchema := pricedSignedManifest()
	badInputSchema["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)["input_schema"] = []any{}
	inputVerification, err := invariants.VerifySignedManifest(badInputSchema)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for input schema: %v", err)
	}
	if inputVerification.StructureValid {
		t.Fatalf("array input_schema must be structurally invalid")
	}

	badOutputSchema := pricedSignedManifest()
	badOutputSchema["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)["output_schema"] = "not an object"
	outputVerification, err := invariants.VerifySignedManifest(badOutputSchema)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for output schema: %v", err)
	}
	if outputVerification.StructureValid {
		t.Fatalf("string output_schema must be structurally invalid")
	}
}

func TestManifestStructureRejectsMalformedPricingMetadata(t *testing.T) {
	cases := []struct {
		label   string
		pricing map[string]any
	}{
		{
			"per_invocation missing unit_price",
			map[string]any{"pricing_model": "per_invocation", "billing_unit": "invocation"},
		},
		{
			"per_unit missing billing_unit",
			map[string]any{
				"pricing_model": "per_unit",
				"unit_price": map[string]any{
					"units":    float64(25),
					"currency": "USD",
				},
			},
		},
		{
			"hybrid missing base_price",
			map[string]any{
				"pricing_model": "hybrid",
				"unit_price": map[string]any{
					"units":    float64(25),
					"currency": "USD",
				},
				"billing_unit": "document",
			},
		},
		{
			"padded billing_unit",
			map[string]any{
				"pricing_model": "per_invocation",
				"unit_price": map[string]any{
					"units":    float64(25),
					"currency": "USD",
				},
				"billing_unit": " invocation",
			},
		},
		{
			"invalid currency",
			map[string]any{
				"pricing_model": "per_invocation",
				"unit_price": map[string]any{
					"units":    float64(25),
					"currency": "usd",
				},
				"billing_unit": "invocation",
			},
		},
		{
			"units above u64 max",
			map[string]any{
				"pricing_model": "per_invocation",
				"unit_price": map[string]any{
					"units":    json.Number("18446744073709551616"),
					"currency": "USD",
				},
				"billing_unit": "invocation",
			},
		},
	}
	for _, tc := range cases {
		t.Run(tc.label, func(t *testing.T) {
			signedManifest := pricedSignedManifest()
			tool := signedManifest["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)
			tool["pricing"] = tc.pricing

			verification, err := invariants.VerifySignedManifest(signedManifest)
			if err != nil {
				t.Fatalf("VerifySignedManifest returned error: %v", err)
			}
			if verification.StructureValid {
				t.Fatalf("%s pricing must be structurally invalid", tc.label)
			}
		})
	}
}

func TestManifestStructureAcceptsRustU64PricingUnits(t *testing.T) {
	signedManifest := pricedSignedManifest()
	tool := signedManifest["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)
	pricing := tool["pricing"].(map[string]any)
	unitPrice := pricing["unit_price"].(map[string]any)
	unitPrice["units"] = json.Number("9223372036854775808")

	verification, err := invariants.VerifySignedManifest(signedManifest)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error: %v", err)
	}
	if !verification.StructureValid {
		t.Fatalf("Rust-valid u64 pricing units above int64 max must be structurally valid")
	}
}

func TestSignedManifestEnvelopeRejectsUnknownTopLevelFields(t *testing.T) {
	signedManifest := pricedSignedManifest()
	signedManifest["unsigned_policy_hint"] = map[string]any{"allow": true}

	verification, err := invariants.VerifySignedManifest(signedManifest)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error: %v", err)
	}
	if verification.StructureValid {
		t.Fatalf("signed manifest envelope with unknown top-level fields must be structurally invalid")
	}
}

func TestManifestStructureAcceptsValidRequiredPermissions(t *testing.T) {
	signedManifest := pricedSignedManifest()
	signedManifest["manifest"].(map[string]any)["required_permissions"] = map[string]any{
		"read_paths":            []any{"/tmp/in"},
		"write_paths":           []any{"/tmp/out"},
		"network_hosts":         []any{"api.example.com"},
		"environment_variables": []any{"TOKEN"},
	}

	verification, err := invariants.VerifySignedManifest(signedManifest)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error: %v", err)
	}
	if !verification.StructureValid {
		t.Fatalf("valid required_permissions must remain structurally valid")
	}
	if verification.SignatureValid {
		t.Fatalf("mutated required_permissions should invalidate the signature")
	}
	if !verification.EmbeddedPublicKeyValid {
		t.Fatalf("embedded public key validity must remain independent")
	}
}

func TestManifestStructureRejectsInvalidRequiredPermissions(t *testing.T) {
	cases := []struct {
		field  string
		values []any
	}{
		{"read_paths", []any{""}},
		{"write_paths", []any{" /tmp/out"}},
		{"network_hosts", []any{"api.example.com "}},
		{"environment_variables", []any{"TOKEN", "TOKEN"}},
		{"read_paths", []any{"/tmp/in\nbad"}},
		{"read_paths", []any{123}},
	}
	for _, tc := range cases {
		t.Run(tc.field, func(t *testing.T) {
			signedManifest := pricedSignedManifest()
			signedManifest["manifest"].(map[string]any)["required_permissions"] = map[string]any{
				tc.field: tc.values,
			}

			verification, err := invariants.VerifySignedManifest(signedManifest)
			if err != nil {
				t.Fatalf("VerifySignedManifest returned error: %v", err)
			}
			if verification.StructureValid {
				t.Fatalf("%s values %#v must be structurally invalid", tc.field, tc.values)
			}
			if verification.SignatureValid {
				t.Fatalf("mutated required_permissions should invalidate the signature")
			}
			if !verification.EmbeddedPublicKeyValid {
				t.Fatalf("embedded public key validity must remain independent")
			}
		})
	}
}

func TestManifestStructureRejectsMalformedRequiredPermissionsObject(t *testing.T) {
	unknownField := pricedSignedManifest()
	unknownField["manifest"].(map[string]any)["required_permissions"] = map[string]any{
		"unknown": []any{"/tmp"},
	}
	unknownVerification, err := invariants.VerifySignedManifest(unknownField)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for unknown field: %v", err)
	}
	if unknownVerification.StructureValid {
		t.Fatalf("unknown required_permissions field must be structurally invalid")
	}

	nonArrayValues := pricedSignedManifest()
	nonArrayValues["manifest"].(map[string]any)["required_permissions"] = map[string]any{
		"read_paths": "/tmp",
	}
	nonArrayVerification, err := invariants.VerifySignedManifest(nonArrayValues)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for non-array value: %v", err)
	}
	if nonArrayVerification.StructureValid {
		t.Fatalf("non-array required_permissions value must be structurally invalid")
	}
}

func TestSignedManifestVerificationIsFailSoftForMalformedEnvelopes(t *testing.T) {
	for _, signedManifest := range []map[string]any{
		{},
		{"manifest": nil, "signature": "", "signer_key": ""},
	} {
		verification, err := invariants.VerifySignedManifest(signedManifest)
		if err != nil {
			t.Fatalf("VerifySignedManifest returned error: %v", err)
		}
		if verification.StructureValid {
			t.Fatalf("malformed envelope must be structurally invalid")
		}
		if verification.SignatureValid {
			t.Fatalf("malformed envelope must not verify signature")
		}
		if verification.EmbeddedPublicKeyValid {
			t.Fatalf("malformed envelope must not report a valid embedded public key")
		}
		if verification.EmbeddedPublicKeyMatchesSigner {
			t.Fatalf("malformed envelope must not report public key match")
		}
	}
}

func TestManifestStructureRejectsMissingRequiredToolFieldsAndUnknownNestedFields(t *testing.T) {
	missingDescription := pricedSignedManifest()
	delete(missingDescription["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any), "description")
	missingDescriptionVerification, err := invariants.VerifySignedManifest(missingDescription)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for missing description: %v", err)
	}
	if missingDescriptionVerification.StructureValid {
		t.Fatalf("missing description must be structurally invalid")
	}

	missingSideEffects := pricedSignedManifest()
	delete(missingSideEffects["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any), "has_side_effects")
	missingSideEffectsVerification, err := invariants.VerifySignedManifest(missingSideEffects)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for missing has_side_effects: %v", err)
	}
	if missingSideEffectsVerification.StructureValid {
		t.Fatalf("missing has_side_effects must be structurally invalid")
	}

	unknownManifestField := pricedSignedManifest()
	unknownManifestField["manifest"].(map[string]any)["unsigned_policy_hint"] = true
	unknownManifestVerification, err := invariants.VerifySignedManifest(unknownManifestField)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for unknown manifest field: %v", err)
	}
	if unknownManifestVerification.StructureValid {
		t.Fatalf("unknown manifest field must be structurally invalid")
	}

	unknownToolField := pricedSignedManifest()
	unknownToolField["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)["annotations"] = map[string]any{}
	unknownToolVerification, err := invariants.VerifySignedManifest(unknownToolField)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for unknown tool field: %v", err)
	}
	if unknownToolVerification.StructureValid {
		t.Fatalf("unknown tool field must be structurally invalid")
	}

	unknownPricingField := pricedSignedManifest()
	unknownPricingField["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)["pricing"].(map[string]any)["experimental_discount"] = 10
	unknownPricingVerification, err := invariants.VerifySignedManifest(unknownPricingField)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for unknown pricing field: %v", err)
	}
	if unknownPricingVerification.StructureValid {
		t.Fatalf("unknown pricing field must be structurally invalid")
	}
}

func TestManifestStructureValidatesServerToolsAndLatencyHintValues(t *testing.T) {
	valid := pricedSignedManifest()
	valid["manifest"].(map[string]any)["server_tools"] = []any{"bash", "text_editor"}
	valid["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)["latency_hint"] = "fast"
	validVerification, err := invariants.VerifySignedManifest(valid)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for valid server_tools: %v", err)
	}
	if !validVerification.StructureValid {
		t.Fatalf("valid server_tools and latency_hint must be structurally valid")
	}

	duplicateServerTool := pricedSignedManifest()
	duplicateServerTool["manifest"].(map[string]any)["server_tools"] = []any{"bash", "bash"}
	duplicateVerification, err := invariants.VerifySignedManifest(duplicateServerTool)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for duplicate server_tools: %v", err)
	}
	if duplicateVerification.StructureValid {
		t.Fatalf("duplicate server_tools must be structurally invalid")
	}

	unknownServerTool := pricedSignedManifest()
	unknownServerTool["manifest"].(map[string]any)["server_tools"] = []any{"database"}
	unknownVerification, err := invariants.VerifySignedManifest(unknownServerTool)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for unknown server_tool: %v", err)
	}
	if unknownVerification.StructureValid {
		t.Fatalf("unknown server_tool must be structurally invalid")
	}

	invalidLatencyHint := pricedSignedManifest()
	invalidLatencyHint["manifest"].(map[string]any)["tools"].([]any)[0].(map[string]any)["latency_hint"] = "immediate"
	invalidLatencyVerification, err := invariants.VerifySignedManifest(invalidLatencyHint)
	if err != nil {
		t.Fatalf("VerifySignedManifest returned error for invalid latency_hint: %v", err)
	}
	if invalidLatencyVerification.StructureValid {
		t.Fatalf("invalid latency_hint must be structurally invalid")
	}
}
